service/storage_proxy.cc

```diff
@@ -3313,8 +3329,16 @@ class data_read_resolver : public abstract_read_resolver {
             }
         }
     }
-    void on_error(gms::inet_address ep, bool disconnect) override {
-        fail_request(read_failure_exception(_schema->ks_name(), _schema->cf_name(), _cl, response_count(), 1, _targets_count, response_count() != 0));
+    void on_error(gms::inet_address ep, error_kind kind) override {
+        switch (kind) {
+        case error_kind::RATE_LIMIT:
+            fail_request(exceptions::rate_limit_exception(_schema->ks_name(), _schema->cf_name()));
```

Comment:
Shouldn't rate limit exception indicate if it was write or read?

Reply:
I'll add an enum indicating the operation type and will include that information in the error description. I also considered different exception types for read rate limit and write rate limit, but I think it will be cumbersome to add more rate limit types using this convention if we wish to do so in the future (e.g. for LWT).

Reply:
Done, added a field which describes the type of the operation. Additionally, I also included information whether an operation was rate limited on the coordinator or not (coordinator rejects are currently guaranteed not to modify the data so this case might be interesting to the users).